core/profiles/src/main/java/software/amazon/awssdk/profiles/ProfileFileSupplier.java

```diff
@@ -95,14 +114,49 @@ static ProfileFileSupplier fixedProfileFile(ProfileFile profileFile) {
     }
 
     /**
-     * creates a {@link ProfileFileSupplier} that produces an existing non-null profile. If the given profile
-     * is null, then the created supplier will also be null.
+     * Creates a {@link ProfileFileSupplier} by combining the {@link ProfileFile} objects from two {@code ProfileFileSupplier}s.
+     * Objects are passed into {@link ProfileFile.Aggregator}.
      *
-     * @param profileFile Profile object to supply.
-     * @return Implementation of {@link ProfileFileSupplier} that is capable of supplying a single profile.
+     * @param suppliers Array of {@code ProfileFileSupplier} objects. {@code ProfileFile} objects are passed to
+     *                  {@link ProfileFile.Aggregator#addFile(ProfileFile)} in the same argument order as the supplier that
+     *                  generated it.
+     * @return Implementation of {@link ProfileFileSupplier} aggregating results from the supplier objects.
      */
-    static ProfileFileSupplier wrapIntoNullableSupplier(ProfileFile profileFile) {
-        return Objects.nonNull(profileFile) ? () -> profileFile : null;
+    static ProfileFileSupplier aggregate(ProfileFileSupplier... suppliers) {
+
+        return new ProfileFileSupplier() {
+
+            volatile ProfileFile currentAggregateProfileFile;
+
+            @Override
+            public ProfileFile get() {
+                ProfileFile.Aggregator aggregator = ProfileFile.aggregator();
+                for (ProfileFileSupplier supplier : suppliers) {
+                    aggregator.addFile(supplier.get());
+                }
+
+                return refreshAndGetCurrentAggregate(aggregator);
+            }
+
+            @Override
+            public void close() {
+                Arrays.stream(suppliers).forEach(ProfileFileSupplier::close);
+            }
+
+            private ProfileFile refreshAndGetCurrentAggregate(ProfileFile.Aggregator aggregator) {
+                ProfileFile next = aggregator.build();
+                if (isNewAggregateProfileFile(next)) {
+                    currentAggregateProfileFile = next;
```

Comment:
Does that need to be thread safe? Do we need to use compare and set?

Reply:
Used same pattern in other classes, for example `ProfileFileRefresher`. Interested to hear what @dagnir has to say, though.

Reply:
Yeah I think we do; we want want to ensure that `next` does not change between lines 148 and 149. Good eye @L-Applin!

If the CAS fails, I think it's fine since that would mean a different thread already updated it.

Reply:
Changed type of reference. Does the `Object.equals()` check break atomicity? I wonder if this needs a synchronized codeblock.